src/agent/agent.c

```diff
@@ -1010,6 +1018,81 @@ static const sd_bus_vtable internal_agent_vtable[] = {
         SD_BUS_VTABLE_END
 };
 
+/*************************************************************************
+ ********** org.containers.hirte.Agent.CreateProxy ***********************
+ ************************************************************************/
+static int agent_method_create_proxy(UNUSED sd_bus_message *m, void *userdata, UNUSED sd_bus_error *ret_error) {
+        Agent *agent = (Agent *) userdata;
+        if (agent == NULL) {
+                hirte_log_error("Failed to create proxy. Agent is NULL.");
+                return EINVAL;
+        }
+
+        _cleanup_sd_bus_message_ sd_bus_message *reply = NULL;
+
+        const char *node_name = NULL;
+        const char *unit_name = NULL;
+        const char *mode = NULL;
+        int r = sd_bus_message_read(m, "sss", &node_name, &unit_name, &mode);
+        if (r < 0) {
+                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_INVALID_ARGS, "Invalid arguments");
+        }
+
+        hirte_log_debugf(
+                        "Received create proxy request for node %s on %s with mode %s",
+                        node_name,
+                        unit_name,
+                        mode);
+
+        _cleanup_proxy_job_ ProxyJob *proxy = proxy_job_new(agent, node_name, unit_name, mode, m);
+        if (proxy == NULL) {
+                return sd_bus_reply_method_errorf(reply, SD_BUS_ERROR_FAILED, "Internal error");
+        }
+
+        if (!proxy_job_export(proxy)) {
+                return sd_bus_reply_method_errorf(reply, SD_BUS_ERROR_FAILED, "Internal error");
+        }
+
+        LIST_APPEND(proxy_jobs, agent->outstanding_proxy_jobs, proxy_job_ref(proxy));
+
+        r = proxy_job_emit_proxy_new(proxy);
+        if (r < 0) {
+                hirte_log_errorf("Failed to emit ProxyNew signal: %s", strerror(-r));
+                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_FAILED, "Internal error");
+        }
+
+        return 1;
+}
+
+/*************************************************************************
+ ********** org.containers.hirte.Agent.RemoveProxy ***********************
+ ************************************************************************/
+static int agent_method_remove_proxy(
+                UNUSED sd_bus_message *m, UNUSED void *userdata, UNUSED sd_bus_error *ret_error) {
+        hirte_log_info("Received remove proxy request");
+
+        /*
+         TODO: Currently empty, has to implement:
```

Comment:
Do we need RemoveProxy? I'm not sure.
We should probably track the lifetime of the client that called AddProxy() and remove the proxy when it dies though.

Reply:
What about the case where the requiring service doesn't need the service on the other node anymore? Shouldn't we clean-up then as well?

Reply:
I was wondering if we could keep the proxy binary alive and keep passing the unit state updates from the manager to the agent and proxy - and the proxy would use [sd_bus_notify](https://www.freedesktop.org/software/systemd/man/sd_notify.html) to notify the systemd service manager. Then we would be able to remove the proxy in both cases, I think. Maybe this is too simplified, though. It would also mean that more resources are required permanently to keep the proxy binary running, not only on starting it. 
What do you think? @alexlarsson

Reply:
So, when I designed this initially I deliberately made the proxy service exit after successful start, because that way we won't have lots of poxy service processes around wasting resources.

However, you're right, and we probably also want to know when the proxy is not needed anymore. I think we can do this by setting `StopWhenUnneeded=yes`, and adding an `ExecStop=` command, which calls this RemoveProxy method. And then yes, we *do* need it.